[driver/sourcekit] Avoid creating temporary output files in TMPDIR

benlangmuir opened the pull request:
When producing frontend arguments for sourcekitd, force the output mode
to -typecheck so that we do not create any temporary output files in the
driver. Previously, any sourcekitd operation that created a compiler
invocation would create 0-sized .o file inside $TMPDIR that would never
be cleaned up.

The new swift-driver project handles temporaries much better as
VirtualPath, and should not need this approach.

rdar://62366123

benlangmuir committed:
[driver/sourcekit] Avoid creating temporary output files in TMPDIR

When producing frontend arguments for sourcekitd, force the output mode
to -typecheck so that we do not create any temporary output files in the
driver. Previously, any sourcekitd operation that created a compiler
invocation would create 0-sized .o file inside $TMPDIR that would never
be cleaned up.

The new swift-driver project handles temporaries much better as
VirtualPath, and should not need this approach.

rdar://62366123

benlangmuir commented:
I still want to add some tests with `swift-ide-test -test-createCompilerInvocation`, but putting up now for feedback on the general approach.  I'm not really thrilled with this, but I wasn't sure how else to avoid creating the temporaries.

benlangmuir committed:
[driver] Add test for new behaviour in createCompilerInvocation

Test for "ForceNoOutputs" flag.

benlangmuir commented:
@swift-ci please smoke test

nathawes commented on lib/Driver/FrontendUtil.cpp:
```diff
@@ -57,6 +58,25 @@ bool swift::driver::getSingleFrontendInvocationFromDriverArguments(
+    ArgList->eraseArg(options::OPT_emit_module_interface);
+    ArgList->eraseArg(options::OPT_emit_module_interface_path);
+    ArgList->eraseArg(options::OPT_emit_objc_header);
+    ArgList->eraseArg(options::OPT_emit_objc_header_path);
```
Does emit_module_doc belong in this group? Or is -typecheck fine with that?

benlangmuir replied:
This is filtering the driver options and `-emit-module-doc-path` is a frontend-only option. My expectation is that removing `-emit-module` will handle both.  I check this in the driver/createCompilerInvocation test case (which is testing the resulting frontend options).

If you think of any other flags I'm missing though, that would be great - I did an audit to look for compatibility issues, but may have missed something.  That's the principle danger in this PR.

benlangmuir commented:
@swift-ci please smoke test Linux

benlangmuir commented:
@brentdax just to clarify I was hoping you could review this as well (reviewers && instead of ||).

brentdax commented on lib/Driver/FrontendUtil.cpp:
```diff
@@ -57,6 +58,25 @@ bool swift::driver::getSingleFrontendInvocationFromDriverArguments(
+  if (ForceNoOutputs) {
+    // Clear existing output modes.
+    ArgList->eraseArg(options::OPT_modes_Group);
+    // Disable other output options that conflict with -typecheck.
```
Nit: There's nothing really "conflicting" here—invoking with `-typecheck` is not supposed to disable these options, we rely on `-typecheck` working this way in many tests, and (e.g.) `-typecheck -emit-module-interface` is the correct way to emit a module interface without any other files. You're disabling them because you don't want any outputs, even secondary ones.

benlangmuir replied:
Hmm, I guess it's not specifically the `-typecheck` flag that they're conflicting with, but typecheck being the only mode/action.  This comment is specific to the `-emit-*-path` options, since once you remove `-emit-module` (via removing the modes group), you get an error if you have `-emit-module-path`, which is why I was describing a conflict:

```
error: this mode does not support emitting modules
```

What about this wording: "Disable -emit-*path options that are invalid after removing -emit-* above" ?

brentdax commented on lib/Driver/FrontendUtil.cpp:
```diff
@@ -57,6 +58,25 @@ bool swift::driver::getSingleFrontendInvocationFromDriverArguments(
+    ArgList->eraseArg(options::OPT_emit_module_interface);
+    ArgList->eraseArg(options::OPT_emit_module_interface_path);
+    ArgList->eraseArg(options::OPT_emit_objc_header);
+    ArgList->eraseArg(options::OPT_emit_objc_header_path);
```
More you might want to disable (plus corresponding `_path`s):

* OPT_emit_dependencies
* OPT_emit_loaded_module_trace
* OPT_emit_private_module_interface_path
* OPT_emit_tbd

There's also OPT_serialize_diagnostics, but I'm guessing you may not want to touch that one.

-----

I worry that this list won't be kept up to date as people add new entries to Options.td. Would it make sense to make a new OptionFlag for "options that cause supplementary outputs to be emitted"? Then you could erase everything from that category (perhaps with some exceptions, like -serialize-diagnostics) and people would just need to appropriately annotate entries when they added them to the options table.

benlangmuir replied:
> There's also OPT_serialize_diagnostics, but I'm guessing you may not want to touch that one.

Why not? We generally shouldn't be serializing diagnostics to a file in sourcekitd.

> Would it make sense to make a new OptionFlag for "options that cause supplementary outputs to be emitted"?

Good idea!

beccadax reviewed (changes requested):
Sorry for the delay, Ben—I didn't see your request.

If you don't want to take action on the comments I've added, this otherwise LGTM.

benlangmuir committed:
[driver] Improve handling of supplementary outputs in createCompilerInvocation

Mark the relevant options in Options.td rather than hard-coding the
list in code, and handle a few more options suggested during the review.

benlangmuir commented:
@swift-ci please test

beccadax reviewed (approved):
Wonderful—thank you!
